File: python/kms-envelope-encrypt.py
import argparse
import sys
from base64 import b64decode, b64encode

import boto3
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad

no_stdin = sys.stdin.isatty()

parser = argparse.ArgumentParser()
parser.add_argument('--key', '-k', help="KMS Data Key to use to decrypt (base64)", required=True)
parser.add_argument('--plaintext', '-p', help="Plaintext to encrypt. Will attempt to read from stdin", default=(None if no_stdin else sys.stdin), required=no_stdin)
args = parser.parse_args()

key = args.k.encode()
plaintext = args.p.encode()

client = boto3.client('kms')

response = client.decrypt(
    CiphertextBlob=b64decode(key)
)
kms_key = response["Plaintext"]

iv = get_random_bytes(16)
cipher = AES.new(kms_key, AES.MODE_CBC, iv=iv)

encrypted_value = cipher.encrypt(pad(plaintext, 16))
final_value = iv + encrypted_value

output = b64encode(final_value)
print(output)

File: node/kmsEnvelopeEncrypt.js
const crypto = require('crypto');
const { KMSClient, DecryptCommand } = require('@aws-sdk/client-kms');
const meow = require('meow');
const getStdin = require('get-stdin');

(async () => {
  const noStdin = process.stdin.isTTY;

  const cli = meow({
    flags: {
      key: {
        alias: 'k',
        isRequired: true,
      },
      plaintext: {
        alias: 'p',
        isRequired: noStdin,
        default: noStdin ? undefined : await getStdin(),
      },
    },
  });

  const { key, plaintext } = cli.flags;

  const client = new KMSClient();

  const kmsKey = await client
    .send(
      new DecryptCommand({
        CiphertextBlob: Buffer.from(key, 'base64'),
      })
    )
    .then((r) => Buffer.from(r.Plaintext));

  const iv = crypto.randomBytes(16);

  const cipher = crypto.createCipheriv('AES-256-CBC', kmsKey, iv);
  cipher.update(plaintext);
  const encryptedData = cipher.final();

  const output = Buffer.concat([iv, encryptedData]).toString('base64');
  console.log(output);
})();
